src/Servers/Kestrel/test/InMemory.FunctionalTests/HttpsTests.cs

```diff
@@ -380,6 +380,42 @@ public async Task ClientAttemptingToUseUnsupportedProtocolIsLoggedAsDebug()
             Assert.Equal(LogLevel.Debug, loggerProvider.FilterLogger.LastLogLevel);
         }
 
+        [Fact]
+        public async Task OnAuthenticate_SeesOtherSettings()
+        {
+            var loggerProvider = new HandshakeErrorLoggerProvider();
+            LoggerFactory.AddProvider(loggerProvider);
+
+            var testCert = TestResources.GetTestCertificate();
+            var onAuthenticateCalled = false;
+
+            await using (var server = new TestServer(context => Task.CompletedTask,
+                new TestServiceContext(LoggerFactory),
+                listenOptions =>
+                {
+                    listenOptions.UseHttps(httpsOptions =>
+                    {
+                        httpsOptions.ServerCertificate = testCert;
+                        httpsOptions.OnAuthenticate = (connectionContext, authOptions) =>
+                        {
+                            Assert.Same(testCert, authOptions.ServerCertificate);
+                            onAuthenticateCalled = true;
```

Comment:
Would it work if you didn't set `httpsOptions.ServerCertificate` and just set in in OnAuthenticate? Can we test that?

Reply:
Almost. We have argument validation on HttpsConnectionAdapter that says you must provide ServerCertificate or ServerCertificateSelector. I can make a test work.